### PR TITLE
Pin zheng improve search experience on dev portal

### PR DESCRIPTION
**Pull request by Ducksss**

### CODE REVIEW

**Comment by Ducksss** [comment]
# Description of change (DP-648)
Based on the specified Figma design and acceptance criteria, I have revamped the search page to improve the search experience. Changes include allowing users to filter for specific information based on their categories, such as products, events and communities. Additional changes include supporting pagination that's only for the desktop view while infinite scrolling on mobile devices. Sub note, changing the filter will reset the current page back to 1.

# Changelogs
- _includes/sidenav-single-level-mobile.html (Formatting)
- _includes/top-nav.html (Added logic to help with state persistence)
- apps/src/search/SearchFullSite.vue (Revamped the entire UI while adding additional logic)
- search/pages.json (Added a sub meta field, named sub_collection to associate a page with a category)
- search/search.html (Modified the layout of the page)

# Images of the changes
(Revamped search page, search params, "Test")
Screenshot 2022-05-09 at 5 19 42 PM

(Revamped search page with filters applied, search params, "Test")
Screenshot 2022-05-09 at 5 20 02 PM

(Revamped search page with filters applied, search params changed to "Personalise")
Screenshot 2022-05-09 at 5 21 11 PM

(Revamped search page with filters applied in mobile view, search params changed to "Personalise")
Screenshot 2022-05-09 at 5 23 02 PM

File: _includes/top-nav.html [hunk 1]
```diff
@@ -51,3 +50,29 @@
...
+        })
+        // Create a query string from the object
+        var queryString = Object.keys(formData).map(key => key + '=' + formData[key]).join('&')
+
+        if (categories) {
+            queryString += '&categories=' + categories
+        }
+
+        // Redirect to the search page with the query string
+        console.log(formData)
```
Comment (teochengyong) [inline _includes/top-nav.html:73]: @Ducksss let's remove the console log

File: _includes/top-nav.html [hunk 2]
```diff
@@ -51,3 +50,29 @@
...
     </nav>
 </div>
+<script>
+    function redirect() {
+        // Create an object to store the values
+        const formData = {}
+        // Getting category from the url
+        const categories = new URL(window.location.href).searchParams.get('categories')
+        // Loop through the inputs and add the values to the object
+        document.querySelectorAll('[name^="query"]').forEach(input => {
```
Comment (teochengyong) [inline _includes/top-nav.html:60]: you can also consider https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams to construct the query strings

File: apps/src/search/SearchFullSite.vue [hunk 3]
```diff
@@ -1,29 +1,236 @@
...
-        }}
-      </h2>
-      <h6 class="is-danger" v-if="errorMsg">{{ errorMsg }}</h6>
-      <div v-if="isNonEmptySearch" class="card-grid-container grid-25rem">
+  <div class="row sidenav-container">
+    <div class="col is-3 is-hidden-touch sidenav">
+      <aside class="sgds-menu">
+        <ul class="sgds-menu-list">
+          <li>
+            <a class="single-level-nav">
```
Comment (teochengyong) [inline apps/src/search/SearchFullSite.vue:7]: Maybe can make this into a component which can be reused throughout the template?

File: apps/src/search/SearchFullSite.vue [hunk 4]
```diff
@@ -1,29 +1,236 @@
...
+                  v-model="categoryCheckbox"
+                />
+                <label
+                  for="checkbox-1"
+                  class="checkbox"
+                  style="flex-grow: 1;"
+                  name="selected-features"
+                  >Products</label
+                >
+              </div>
```
Comment (teochengyong) [inline apps/src/search/SearchFullSite.vue:122]: @Ducksss I think we can make this into a component

File: _includes/top-nav.html (hunk 1 above)
Comment (Ducksss) [reply _includes/top-nav.html:73]: okie, will do

File: apps/src/search/SearchFullSite.vue (hunk 4 above)
Comment (Ducksss) [reply apps/src/search/SearchFullSite.vue:122]: ok! sure, will do :)

**Comment by yanbutan** [comment]
Heyoh @Ducksss I haven't properly started review but noticed sth.

https://user-images.githubusercontent.com/44741317/171156548-4ddaa654-9b35-4a7c-8354-9463f4259d1b.mov

so after i unchecked the categories, and i click back button on browser, it changes back to previous urls which were previous category selections. I found this a bit weird because it doesn't refresh the page with the categories again nor does it bring the user back to previous page before this search page. So is just awkwardly going back through the url hahah.

Any ideas on how to make this better? I reckon you should route user back to previous page (the page they visit search from) when they click browser back because to me I wont need to go back to see what were my category selections, I will just select them from the filter on the page. Not too sure how to achieve this tho.

@teochengyong any thoughts?

**Comment by yanbutan** [comment]
@Ducksss hello sir i tried out composition api you can check it out at improved-search-composition branch. Change was pretty smooth but I got stuck at a different issue. There's some jankiness with back navigation, i'm not sure if you already know. If you already know, could I ask how are you handling it? it's quite unintuative now. 

Anyways maybe can call tmr to explain what i mean and if you have qns bah.

**Comment by yanbutan** [comment]
@Ducksss hello sir i tried out composition api you can check it out at improved-search-composition branch. Change was pretty smooth but I got stuck at a different issue. There's some jankiness with back navigation, i'm not sure if you already know. If you already know, could I ask how are you handling it? it's quite unintuative now. 

Anyways maybe can call tmr to explain what i mean and if you have qns bah.

**Pull request closed without merging**